File: 101. Symmetric Tree/isSymmetric.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */

var isSymmetric = function(root) {
    var isSym = (L,R) => {
        if (L && R && L.val===R.val) {
            return isSym(L.left,R.right) && isSym(L.right,R.left)
        }
        else {
            return L===R
        }
    }
    
    if (root) {
        return isSym(root.left,root.right)
    }
    else {
        return true
    }
};

// Runtime 128 ms / Memory 44.1 MB
// debug: https://jsfiddle.net/

File: 101. Symmetric Tree/isSymmetric.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        return check(root->left,root->right);
    }

    bool check(TreeNode* p, TreeNode* q) {
        // goto leaf
        if (!p && !q) { return true; }
        else if (!p || !q) { return false; }
        // check
        if (p->val != q->val) { return false; }
        else {
            bool L = check(p->left,q->right); //mirror
            bool R = check(p->right,q->left); //mirror
            return L && R;
        }
    }
};